jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpChannelOverHTTP2.java

```diff
@@ -261,31 +268,408 @@ public InvocationType getInvocationType()
             {
                 return callback.getInvocationType();
             }
-        });
+        };
+        _contentDemander.onContent(content);
+        boolean handle = onContent(content);
 
         boolean endStream = frame.isEndStream();
         if (endStream)
         {
             boolean handleContent = onContentComplete();
+            // this will generate EOF -> must happen before wakeup
             boolean handleRequest = onRequestComplete();
             handle |= handleContent | handleRequest;
         }
 
+        boolean woken = getRequest().getHttpInput().wakeup();
         if (LOG.isDebugEnabled())
         {
-            LOG.debug("HTTP2 Request #{}/{}: {} bytes of {} content, handle: {}",
+            Stream stream = getStream();
+            LOG.debug("HTTP2 Request #{}/{}: {} bytes of {} content, woken: {}, handle: {}",
                     stream.getId(),
                     Integer.toHexString(stream.getSession().hashCode()),
                     length,
                     endStream ? "last" : "some",
+                    woken,
                     handle);
         }
+        handle |= woken;
 
         boolean wasDelayed = _delayedUntilContent;
         _delayedUntilContent = false;
         return handle || wasDelayed ? this : null;
     }
 
+    private interface ContentDemander
```

Comment:
I don't think we need both implementations now.  Let's just go for the atomic one, so we can probably get rid of the interface.

Reply:
Done.